core/server/data/migrations/utils.js

```diff
@@ -0,0 +1,244 @@
+// @ts-check
+const ObjectId = require('bson-objectid').default;
+const logging = require('../../../shared/logging');
+
+/**
+ * Creates a migration which will add a permission to the database
+ *
+ * @param {Object} config
+ * @param {string} config.name - The name of the permission
+ * @param {string} config.action - The action_type of the permission
+ * @param {string} config.object - The object_type of the permission
+ *
+ * @returns {Migration}
+ */
+function addPermission(config) {
+    return createTransactionalMigration(
+        async function up(connection) {
+            const existingPermission = await connection('permissions').where({
+                name: config.name,
+                action_type: config.action,
+                object_type: config.object
+            }).first();
+
+            if (existingPermission) {
+                logging.warn(`Permission for ${config.action}:${config.object} already added`);
+                return;
+            }
+
+            logging.info(`Adding permission for ${config.action}:${config.object}`);
+
+            const date = connection.raw('CURRENT_TIMESTAMP');
+
+            await connection('permissions').insert({
+                id: ObjectId.generate(),
+                name: config.name,
+                action_type: config.action,
+                object_type: config.object,
+                created_at: date,
+                created_by: 1,
+                updated_at: date,
+                updated_by: 1
+            });
+        },
+        async function down(connection) {
+            const existingPermission = await connection('permissions').where({
+                name: config.name,
+                action_type: config.action,
+                object_type: config.object
+            }).first();
+
+            if (!existingPermission) {
+                logging.warn(`Permission for ${config.action}:${config.object} already removed`);
+                return;
+            }
+
+            logging.info(`Removing permission for ${config.action}:${config.object}`);
+
+            await connection('permissions').where({
+                action_type: config.action,
+                object_type: config.object
+            }).del();
+        }
+    );
+}
+
+/**
+ * Creates a migration which will link a permission to a role in the database
+ *
+ * @param {Object} config
+ * @param {string} config.permission - The name of the permission
+ * @param {string} config.role - The name of the role
+ *
+ * @returns {Migration}
+ */
+function addPermissionToRole(config) {
+    return createTransactionalMigration(
+        async function up(connection) {
+            const permission = await connection('permissions').where({
+                name: config.permission
+            }).first();
+
+            if (!permission) {
+                throw new Error(
+                    `Cannot add perm(${config.permission}) to role(${config.role}) - perm does not exist`
```

Comment:
Can we always use the word "permission" a perm is a hairdo 😝

Reply:
:person_curly_hair: Fine!!! :sweat_smile: